### PR TITLE
Mark the Flutter Views as focusable by setting a tabindex value.

### PR DESCRIPTION
**Pull request by tugorez**
Mark the Flutter View as focusable by setting a tabindex value.

* When a given flutter view is focused its tabindex will be `-1`
* When a given flutter view is not focused its tabindex will be `0`
* When semantics are enabled no tabindex will be set.

Relevant Issues are:

* Design doc: https://flutter.dev/go/focus-management 
* Focus in web multiview: https://github.com/flutter/flutter/issues/137443
* Platform dispatcher changes: https://github.com/flutter/engine/pull/49841

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/window.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/window.dart]: Unused method? Delete?

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:79]: What if `lastViewId` is null? (That's why you flipped this if/else, right? ;) )

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher.dart:231]:
```suggestion
  final ViewFocusBinding _viewFocusBinding = ViewFocusBinding(
```

It is not `late` when you're initializing it immediately, is it?

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:87]: WHAT IF...

The `ViewFocusBinding` class exposed a:

```
Stream<ViewFocusEvent> onFocusChanged;
```

And users from outside (say, the PlatformDispatcher) could subscribe to that stream, as needed?

Internally, you could have a StreamController that lets you check when the stream is listened/unlistened to, and then attach/detach the DOM events as needed (similar to what you had before with the array of listeners!)

That way, this class doesn't need to keep a copy of the `_onViewFocusChange` handler, AND potentially multiple classes could listen to the ViewFocusChange events for their own purposes (say one for the framework, and another for debugging)

((PS: This also solves the problem that we have with the `KeyboardBinding` not exposing the events that we might need elsewhere in the engine, and instead being very rigid in what their callback/handling of events is))

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher.dart:233]: If this class exposed a stream of events, then you don't need to pass the `onViewFocusChange` through its constructor, and the `viewManager` can default to `EnginePlatformDispatcher.instance.viewManager`, so `ViewFocusBinding` can become a singleton again :P

File: lib/web_ui/lib/src/engine/view_embedder/style_manager.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/view_embedder/style_manager.dart:121]: With the current code, is this even visible? Do we ever leave the focus in the flutter-view root?

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:79]: Good catch. Reverted the if/else.

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:231]: It is late because I need the view manager (which is not static).

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:87]: Great advice @ditman. What a true pro! 

Done.

File: lib/web_ui/lib/src/engine/view_embedder/style_manager.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/view_embedder/style_manager.dart:121]: > With the current code, is this even visible?
Yes it is

> Do we ever leave the focus in the flutter-view root?
Yes, when the widget traversal is done the idea is that we should jump to the next/previous DOM node.

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:233]: Hmmmm I find a bit weird the fact that EnginePlatformDispatcher depends on ViewFocusBinding and viceversa.

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:96]: Nothing is ever free :P. 

We'd have to be careful with memory as that map will have an entry per unique ever focused DOMElement.

Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:96]: NVM, got your point! Done.

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart]: ? It already depends on it, I'm just saying that instead of:

```dart
final _viewFocusBinding = ViewFocusBinding(
  viewManager: viewManager,
  onViewFocusChange: invokeOnViewFocusChange,
)
```

you do:

```dart
final _viewFocusBinding = ViewFocusBinding(
  viewManager: viewManager,
); // or: ViewFocusBinding.instance;

// elsewhere

_viewFocusBinding.onViewFocusChange.listen((ViewFocusChange event) {
  // Do whatever we need with the event now...
});
```

That way you don't need to pass a tear-off into the ViewFocusBinding instance?

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher.dart:82]: I want to refactor this to spit out a stream now... :/ :P

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:20]: `sync: true` may have exciting consequences, according to the docs :P

Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:20]: Also: you can use `onListen` and `onCancel` on the `StreamController` to know if anybody is listening to the stream. You could use those events to call the code that you have in `_init` as late as needed and in `dispose` as early as possible.

The actual implementation of `dispose` could be: `_onViewFocusChangeController.close()`? (which should cause all the attached streams to cancel, if any?)

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:91]: Agh, I thought we were setting the focusin/focusout listeners in the rootElement 🤦

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart]: > ? It already depends on it, I'm just saying that instead of:

How so? The view focus binding class  doesn't have any dependencies on `EnginePlatformDispatcher` only on `FlutterViewFocusManager`

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:82]: > I want to refactor this to spit out a stream now... :/ :P

What would be the advante of doing something like that?

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:91]: Yeah, easy to miss; but no, all those events are attached to the body.

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:20]: > sync: true may have exciting consequences, according to the docs :P

Great call removed.

> Also: you can use onListen and onCancel on the StreamController to know if anybody is listening to the stream. You could use those events to call the code that you have in _init as late as needed and in dispose as early as possible.

I still prefer the existing implementation (realized that the early we listen for these focus changes, the better). 
Are there any flaws with it? Happy to change if you feel strongly about it.

File: lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart
Comment (tugorez) [reply lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart:146]: Any advice on where it should go?

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:20]: We normally use callbacks rather than streams, and callbacks are synchronous. So I wouldn't say that using a synchronous stream would be non-idiomatic. In fact, I'd say that using streams in the first place would be non-idiomatic, whether they are sync or not :)

I think this depends on what semantics we actually want here. Sync vs async will dictate the timing, potentially race conditions.

**Review by yjbanov — APPROVED** [review]
LGTM if LGT @ditman

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:20]: Ah yes, the problem with tests is that we're not waiting for the event to propagate and have the listener do its thing. `sync` works in this case because it makes the callback sync, and also is valid because it's just propagating the async-ness of a browser event.

This could be made fully async. I suspect tests would be fixed just by adding a listener on the same stream, awaiting for the next event to happen (`await blah.stream.first`), and THEN checking the DOM right after :)

(We just need that signal to know that the DOM has been modified)

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:82]: Not passing single-purpose tear-offs around, and letting disjoint parts of the application listen to the same events. The same as why "addEventListener('click')" is better than setting "onclick" :P

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart]: > The view focus binding class doesn't have any dependencies on `EnginePlatformDispatcher` only on `FlutterViewFocusManager`

Could you create the view focus binding class outside of the `EnginePlatformDispatcher`, without using the `EnginePlatformDispatcher` (to retrieve the `viewManager`? I think that's a fairly strong dependency :P

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (ditman) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:20]: Having said this, it's true that the callback is simple, and there's no clear win on why this should emit a stream, so YAGNI says: do the callback.

Apologies for the churn!

Comment (tugorez) [reply lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart]: Nothing to apologize for @ditman ! Getting the right design abstractions is hard and I'm glad we chatted about different options and its trade-offs!

**Comment by ditman** [comment]
(Holding off on reviewing this, @tugorez might want to tweak some odds and ends, will follow up with him next week! Thanks for the feature @tugorez!!)

**Comment by tugorez** [comment]
https://flutter-web-20c94.web.app/

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:33]: Can we attach focusin/focusout to the rootElements of the registered views? Wouldn't that remove a lot of the gymnastics we need to do to see if the focus event comes from a flutter view or not? (the answer will always be: "yes")

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:84]: Basically it'd guarantee in this case that every `element` would have a parent (or itself) with a flutter view id, so we can make this an `int _viewId` function, no?

File: lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart
Comment (ditman) [inline lib/web_ui/lib/src/engine/platform_dispatcher/view_focus_binding.dart:93]: If we were to listen to focusin/out, we'd need to use this method to attach listener to new views.

**Review by ditman — APPROVED** [review]
This LGTM, but I'd like the focusin/out events to be attached to the rootElements of the views, if technically possible. I think that'd simplify this code a good bit!

**Comment by tugorez** [comment]
I made some changes.

- App with engine without these changes: https://a11ybefore.web.app/
- App with engine with these changes: https://a11yafter.web.app/

**Comment by tugorez** [comment]
> @tugorez you're still testing with platform views, right?

Waiting on a final extra review from @yjbanov to make sure my changes don't break semantics.

**Comment by ditman** [comment]
> I made some changes.
> 
> * App with engine without these changes: https://a11ybefore.web.app/
> * App with engine with these changes: https://a11yafter.web.app/

Took a look with VoiceOver, and both apps seem to work the same for me. Let's merge this!

**Comment by tugorez** [comment]
Let's land this PR as is and follow up with the multiview verification in https://github.com/flutter/flutter/issues/144853

File: lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart
Comment (tugorez) [reply lib/web_ui/test/engine/platform_dispatcher/view_focus_binding_test.dart:170]: Happy to address this in a follow up PR!

**Comment by ditman** [comment]
Let's go @tugorez!!

**Pull request merged**
